### PR TITLE
feat: add experimental flag to sam local invoke, and sam local start-lambda commands 

### PR DESCRIPTION
**Pull request by moelasmar**
This PR is about adding the beta feature flag to `invoke` and `start-lambda` sam local commands. So, in case of the usage of Terraform support, the customer has to provide the beta feature flag, either by adding `--beta-features` option, or SAM CLI will prompt the customer asking for approval to run the command with the beta feature flag.

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/_utils/options.py [hunk 1]
```diff
@@ -37,43 +37,6 @@
 LOG = logging.getLogger(__name__)
 
 
-def parameterized_option(option):
```
Comment (moelasmar) [inline samcli/commands/_utils/options.py:40]: moved to a new module to avoid import cycles.

**Review by hawflau — APPROVED** [review]

File: samcli/commands/_utils/custom_options/hook_package_id_option.py [hunk 2]
```diff
@@ -46,6 +50,20 @@ def handle_parse_result(self, ctx, opts, args):
...
+                terraform_support_message = (
+                    "Supporting Terraform applications is a beta feature.\n"
+                    "Please confirm if you would like to proceed using SAM CLI with terraform application.\n"
+                    f"You can also enable this beta feature with 'sam {command_name} --beta-features'."
+                )
+                if not prompt_experimental(ExperimentalFlag.TerraformSupport, terraform_support_message):
+                    LOG.debug("Experimental flag is disabled, return, and do not run the prepare hook")
+                    return super().handle_parse_result(ctx, opts, args)
+            elif not beta_features:
+                LOG.debug("beta-feature flag is disabled, return, and do not run the prepare hook")
```
Comment (sriram-mv) [inline samcli/commands/_utils/custom_options/hook_package_id_option.py:65]: `beta-feature flag is disabled and prepare hook is not run`

File: samcli/commands/_utils/custom_options/hook_package_id_option.py [hunk 3]
```diff
@@ -46,6 +50,20 @@ def handle_parse_result(self, ctx, opts, args):
...
+            # check beta-feature
+            beta_features = opts.get("beta_features")
+            if beta_features is None:
+                terraform_support_message = (
+                    "Supporting Terraform applications is a beta feature.\n"
+                    "Please confirm if you would like to proceed using SAM CLI with terraform application.\n"
+                    f"You can also enable this beta feature with 'sam {command_name} --beta-features'."
+                )
+                if not prompt_experimental(ExperimentalFlag.TerraformSupport, terraform_support_message):
+                    LOG.debug("Experimental flag is disabled, return, and do not run the prepare hook")
```
Comment (sriram-mv) [inline samcli/commands/_utils/custom_options/hook_package_id_option.py:62]: Experimental flag is disabled and prepare hook is not run.

File: samcli/commands/_utils/custom_options/hook_package_id_option.py [hunk 4]
```diff
@@ -46,6 +50,20 @@ def handle_parse_result(self, ctx, opts, args):
                     raise click.BadParameter(
                         f"Parameters {opt_name}, and {','.join(self._invalid_coexist_options)} can not be used together"
                     )
+            # check beta-feature
+            beta_features = opts.get("beta_features")
+            if beta_features is None:
+                terraform_support_message = (
+                    "Supporting Terraform applications is a beta feature.\n"
```
Comment (sriram-mv) [inline samcli/commands/_utils/custom_options/hook_package_id_option.py:57]: hook option are a general IaC prepare hook right? we may need a way to figure out its terraform features that are being enabled and not some other Iac?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: samcli/commands/_utils/custom_options/hook_package_id_option.py (hunk 4 above)
Comment (moelasmar) [reply samcli/commands/_utils/custom_options/hook_package_id_option.py:57]: for now, we only have Terraform supported, and we are going to go GA and remove beta features flag before adding another IaC support. I think we can handle this scenario when we add a new IaC support, or start accepting external hook packages.

File: samcli/commands/_utils/custom_options/hook_package_id_option.py [hunk 5]
```diff
@@ -52,17 +52,20 @@ def handle_parse_result(self, ctx, opts, args):
...
-                    LOG.debug("Experimental flag is disabled, return, and do not run the prepare hook")
+
+            # check if beta feature flag is required for a specific hook package
+            # The IaCs support experimental flag map will contain only the beta IaCs. In case we support the external
+            # hooks, we need to first know that the hook package is an external, and to handle the beta feature of it
+            # using different approach
+            experimental_entry = ExperimentalFlag.IaCsSupport.get(hook_package_id)
+            if beta_features is None and experimental_entry is not None:
+
+                terraform_support_message = _get_iac_support_experimental_prompt_message(hook_package_id, command_name)
```
Comment (sriram-mv) [inline samcli/commands/_utils/custom_options/hook_package_id_option.py:63]: nit: rename terraform_support_message to iac_support_message

**Review by sriram-mv — APPROVED** [review]

**Pull request merged**